libs/vkd3d/resource.c

```diff
@@ -6430,47 +6426,92 @@ static void vkd3d_memory_info_get_topology(struct vkd3d_memory_topology *topolog
+    if (vkd3d_config_flags & VKD3D_CONFIG_FLAG_NO_UPLOAD_HVV)
+    {
+        INFO("Topology: Forcing HOST_COHERENT for UPLOAD heap.\n");
+        return VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+    }
 
-    /* Mask out any memory types which are deemed problematic. */
-    for (i = 0, mask = 0; i < device->memory_properties.memoryTypeCount; i++)
+    if (vkd3d_memory_topology_is_uma_like(topology))
     {
-        const VkMemoryPropertyFlags pinned_mask = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
-                                                  VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT;
-        flags = device->memory_properties.memoryTypes[i].propertyFlags;
-        heap_index = device->memory_properties.memoryTypes[i].heapIndex;
+        /* Verify that there exists a DEVICE_LOCAL type that is not HOST_VISIBLE on this device
+         * which maps to the largest device local heap. That way, it is safe to mask out all memory types which are
+         * DEVICE_LOCAL | HOST_VISIBLE.
+         * Similarly, there must exist a host-only type. */
+        INFO("Topology: UMA-like topology. Using DEVICE_LOCAL | HOST_COHERENT for UPLOAD.\n");
+        return VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT |
+                VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
+                VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+    }
+    else if (topology->device_local_heap_count <= 1)
+    {
+        VkDeviceSize largest_size = device->memory_properties.memoryHeaps[topology->largest_device_local_heap_index].size;
+        const VkDeviceSize minimum_rebar_size = 5ull * 1000ull * 1000ull * 1000ull;
 
-        if (heap_index != topology->largest_device_local_heap_index &&
-            heap_index != topology->largest_host_only_heap_index &&
-            (flags & pinned_mask) == pinned_mask)
+        if (largest_size < minimum_rebar_size)
+        {
+            INFO("Topology largest device local heap is too small (%"PRIu64" bytes) for effective ReBAR, using HOST_COHERENT for UPLOAD.\n", largest_size);
+            return VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+        }
+        else
         {
-            mask |= 1u << i;
-            WARN("Blocking memory type %u for use (PCI-pinned memory).\n", i);
+            /* If we only have one device local heap. */
+            INFO("Topology: No more than 1 device local heap, assuming ReBAR-style access. Using DEVICE_LOCAL | HOST_COHERENT for UPLOAD.\n");
+            /* If DEVICE_LOCAL_BIT does not actually exist, that is fine,
+             * we'll fallback to VISIBLE | COHERENT when allocating. */
+            return VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT |
+                    VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
+                    VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
         }
     }
-
-    return ~mask;
+    else
+    {
+        INFO("Topology: Device heaps are split. Assuming small BAR situation. Using DEVICE_LOCAL | HOST_COHERENT.\n");
```

Comment:
Should we have a debug flag (the opposite to `no_upload_hvv`) to force HVV use even on small BAR systems?

Reply:
I don't think so.